README.md
# Chat
Simple webchat. Written in Elm and Haskell. Uses websockets.

## Installing
In `/frontend`:
```bash
npm install chokidar-cli --save-dev
elm-package install
```

In `/backend`:
```bash
cabal install websockets
cd ../out/backend
ghc -o server ../../backend/server.hs -odir . -hidir .
```

## Developing
Start a file watch server:
```bash
npm run watch
```

Copy static files to the output directory:
```bash
cp assets/* out/frontend
```

Run elm-reactor to be able to use it as a server:
```bash
elm-reactor
```

Run backend (`127.0.0.1:9160`):
```bash
./server
```

## Deployment
Configure Nginx (or whatever) to serve frontend statically.

Run `./server` in background process.

## Running docker
```bash
docker build --rm -f Dockerfile -t lambda-chat:latest .
docker run lambda-chat -p 80:80
```

## Explanation

#### What is `elm-package.json` for?
Dependencies of Elm project.

#### What is `package.json` for?
Chokidar is used to watch files and retranspile Elm every time some of its file changes.


backend/ChangeLog.md
# Changelog for lambda-chat-backend

## Unreleased changes


backend/test/Spec.hs
import Test.Hspec

main :: IO ()
main = hspec spec

spec :: Spec
spec = do
  describe "even" $ do
    it "checks if number is even" $ do
      even 3 `shouldBe` False

backend/test/QuickSpec.hs
module QuickSpec where

import Test.QuickCheck

reverse :: [a] -> [a]
reverse []     = []
reverse (x:xs) = Prelude.reverse xs ++ [x]

prop_ReverseReverseId :: [Integer] -> Bool
prop_ReverseReverseId xs = Prelude.reverse (Prelude.reverse xs) == xs

main = quickCheck prop_ReverseReverseId


backend/test/ServerSpec.hs
module ServerSpec where

import Test.Hspec

main = hspec spec

spec = do
  describe "even" $ do
    it "checks if number is even" $
      even 3 `shouldBe` False
      


backend/plain.hs
{-# LANGUAGE OverloadedStrings #-}
module Main where
import Data.Char (isPunctuation, isSpace)
import Data.Monoid (mappend)
import Data.Text (Text)
import Control.Exception (finally)
import Control.Monad (forM_, forever)
import Control.Concurrent (MVar, newMVar, modifyMVar_, modifyMVar, readMVar)
import qualified Data.Text as T
import qualified Data.Text.IO as T

import qualified Network.WebSockets as WS

type Client = (Text, WS.Connection)
type ServerState = [Client]

newServerState :: ServerState
newServerState = []

numClients :: ServerState -> Int
numClients = length

clientExists :: Client -> ServerState -> Bool
clientExists client = any ((== fst client) . fst)

addClient :: Client -> ServerState -> ServerState
addClient client clients = client : clients

removeClient :: Client -> ServerState -> ServerState
removeClient client = filter ((/= fst client) . fst)

broadcast :: Text -> ServerState -> IO ()
broadcast message clients = do
  T.putStrLn message
  forM_ clients $ \(_, conn) -> WS.sendTextData conn message

main :: IO ()
main = do
  state <- newMVar newServerState
  WS.runServer "127.0.0.1" 9160 $ application state

application :: MVar ServerState -> WS.ServerApp
application state pending = do
  conn <- WS.acceptRequest pending
  WS.forkPingThread conn 30
  msg <- WS.receiveData conn
  clients <- readMVar state
  case msg of
    _ | not (prefix `T.isPrefixOf` msg) ->
        WS.sendTextData conn ("Wrong announcement." :: Text)
      | any ($ fst client)
        [T.null, T.any isPunctuation, T.any isSpace] ->
            WS.sendTextData conn ("Name cannot " `mappend`
                "contain punctuation or whitespace, and " `mappend`
                "cannot be empty." :: Text)
      | clientExists client clients ->
        WS.sendTextData conn ("User already exists." :: Text)
      | otherwise -> flip finally disconnect $ do
          modifyMVar_ state $ \s -> do
            let s' = addClient client s
        --    WS.sendTextData conn $
        --      "SYSTEM: " `mappend` T.intercalate "," (map fst s) `mappend` " are online"
            broadcast ("SYSTEM: " `mappend` fst client `mappend` " joined.") s'
            return s'
          talk conn state client
    where
      prefix     = "JOIN "
      client     = (T.drop (T.length prefix) msg, conn)
      disconnect = do
        s <- modifyMVar state $ \s ->
          let s' = removeClient client s in return (s', s')
        broadcast ("SYSTEM: " `mappend` fst client `mappend` " disconnected.") s

talk :: WS.Connection -> MVar ServerState -> Client -> IO ()
talk conn state (user, _) = forever $ do
  msg <- WS.receiveData conn
  readMVar state >>= broadcast
    (user `mappend` ": " `mappend` msg)

backend/server.hs
{-# LANGUAGE OverloadedStrings #-}
module Main where
import Data.Char (isPunctuation, isSpace)
import Data.Monoid (mappend)
import Data.Text (Text)
import Data.String
import Data.List
import Control.Exception (finally)
import Control.Monad (forM_, forever)
import Control.Concurrent (MVar, newMVar, modifyMVar_, modifyMVar, readMVar)
import qualified Data.Text as T
import qualified Data.Text.IO as T
import Text.Regex.Posix
import qualified Network.WebSockets as WS

type Client = (Text, WS.Connection)
type ServerState = [Client]

newServerState :: ServerState
newServerState = []

numClients :: ServerState -> Int
numClients = length

clientExists :: Client -> ServerState -> Bool
clientExists client = any ((== fst client) . fst)

addClient :: Client -> ServerState -> ServerState
addClient client clients = client : clients

removeClient :: Client -> ServerState -> ServerState
removeClient client = filter ((/= fst client) . fst)

sendDirectMessage :: Text -> Text -> ServerState -> IO ()
sendDirectMessage user msg clients = do
  let recipient = words (T.unpack msg)!!1
  let client = find (\(name, _)->T.unpack name==recipient) clients

  case client of
    Nothing -> broadcast
      ("SYSTEM: " `mappend` user `mappend` " is extremely stupid. She or he wanna send message to user that does not exist😳") clients
    Just (_, conn) ->
      WS.sendTextData conn (user `mappend` " directly to you: " `mappend` T.pack(drop  (8 + length (T.unpack user)) (T.unpack msg)))

session :: WS.Connection -> MVar ServerState -> Client -> IO()
session conn state client = do
   modifyMVar_ state $ \s -> do
     let s' = addClient client s
     broadcast ("SYSTEM: " `mappend` fst client `mappend` " joined.") s'
     return s'
   talk conn state client


sendMessageWithVulgarism :: MVar ServerState -> Text -> Text -> IO()
sendMessageWithVulgarism state user msg =
  readMVar state >>= broadcast
    (user `mappend`": " `mappend` T.replace "fuck" "f*ck" msg)

broadcast :: Text -> ServerState -> IO ()
broadcast message clients = do
  T.putStrLn message
  forM_ clients $ \(_, conn) -> WS.sendTextData conn message

main :: IO ()
main = do
  state <- newMVar newServerState
  T.putStrLn "I'm server and I'm here to serve you."
  WS.runServer "127.0.0.1" 9160 $ application state

application :: MVar ServerState -> WS.ServerApp
application state pending = do
  conn <- WS.acceptRequest pending
  WS.forkPingThread conn 30
  msg <- WS.receiveData conn
  clients <- readMVar state
  case msg of
    _   | any ($ fst client)
          [T.null, T.any isPunctuation, T.any isSpace] ->
          WS.sendTextData conn ("Name cannot " `mappend`
          "contain punctuation or whitespace, and " `mappend`
          "cannot be empty." :: Text)
        | clientExists client clients ->
          WS.sendTextData conn ("User already exists." :: Text)
        | otherwise -> finally (session conn state client) disconnect

            where
              client     = (T.drop 5 msg, conn)
              disconnect = do
                  s <- modifyMVar state $ \s ->
                      let s' = removeClient client s in return (s', s')
                  broadcast ("SYSTEM: " `mappend` fst client `mappend` " disconnected.") s

talk :: WS.Connection -> MVar ServerState -> Client -> IO ()
talk conn state (user, _) = forever $ do
  msg <- WS.receiveData conn
  case msg of
    _   | ("direct " `T.isPrefixOf` msg) && (length (words $ T.unpack msg) > 4) -> readMVar state >>= broadcast
        ("SYSTEM: " `mappend` user `mappend` " is so stupid that she or he cannot send proper direct message!😄")
        | "direct " `T.isPrefixOf` msg -> readMVar state >>= sendDirectMessage user msg
        | T.toLower msg == T.pack "ping" -> WS.sendTextData conn (T.pack "👻: pong")
        | T.unpack (T.toLower msg) =~ ("fuck" ::String) -> sendMessageWithVulgarism state user msg
        | otherwise -> readMVar state >>= broadcast
          (user `mappend` ": " `mappend` msg)


backend/src/Server.hs
{-# LANGUAGE OverloadedStrings #-}

module Server ( server ) where

import Data.Char (isPunctuation, isSpace)
import Data.Monoid (mappend)
import Data.Text (Text)
import Data.String
import Data.List
import Control.Exception (finally)
import Control.Monad (forM_, forever)
import Control.Concurrent (MVar, newMVar, modifyMVar_, modifyMVar, readMVar)
import qualified Data.Text as T
import qualified Data.Text.IO as T
import Text.Regex.Posix
import qualified Network.WebSockets as WS

type Client = (Text, WS.Connection)
type ServerState = [Client]

newServerState :: ServerState
newServerState = []

numClients :: ServerState -> Int
numClients = length

clientExists :: Client -> ServerState -> Bool
clientExists client = any ((== fst client) . fst)

addClient :: Client -> ServerState -> ServerState
addClient client clients = client : clients

removeClient :: Client -> ServerState -> ServerState
removeClient client = filter ((/= fst client) . fst)

sendDirectMessage :: Text -> Text -> ServerState -> IO ()
sendDirectMessage user msg clients = do
  let recipient = words (T.unpack msg)!!1
  let client = find (\(name, _)->T.unpack name==recipient) clients

  case client of
    Nothing -> broadcast
      ("SYSTEM: " `mappend` user `mappend` " is extremely stupid. She or he wanna send message to user that does not exist😳") clients
    Just (_, conn) ->
      WS.sendTextData conn (user `mappend` " directly to you: " `mappend` T.pack(drop  (8 + length (T.unpack user)) (T.unpack msg)))

session :: WS.Connection -> MVar ServerState -> Client -> IO()
session conn state client = do
  modifyMVar_ state $ \s -> do
    let s' = addClient client s
    broadcast ("SYSTEM: " `mappend` fst client `mappend` " joined.") s'
    return s'
  talk conn state client


sendMessageWithVulgarism :: MVar ServerState -> Text -> Text -> IO()
sendMessageWithVulgarism state user msg =
  readMVar state >>= broadcast
    (user `mappend`": " `mappend` T.replace "fuck" "f*ck" msg)

broadcast :: Text -> ServerState -> IO ()
broadcast message clients = do
  T.putStrLn message
  forM_ clients $ \(_, conn) -> WS.sendTextData conn message

server :: IO ()
server = do
  state <- newMVar newServerState
  T.putStrLn "I'm server and I'm here to serve you."
  WS.runServer "127.0.0.1" 9160 $ application state

application :: MVar ServerState -> WS.ServerApp
application state pending = do
  conn <- WS.acceptRequest pending
  WS.forkPingThread conn 30
  msg <- WS.receiveData conn
  clients <- readMVar state
  case msg of
    _   | any ($ fst client)
          [T.null, T.any isPunctuation, T.any isSpace] ->
          WS.sendTextData conn ("Name cannot " `mappend`
          "contain punctuation or whitespace, and " `mappend`
          "cannot be empty." :: Text)
        | clientExists client clients ->
          WS.sendTextData conn ("User already exists." :: Text)
        | otherwise -> finally (session conn state client) disconnect

            where
              client     = (T.drop 5 msg, conn)
              disconnect = do
                  s <- modifyMVar state $ \s ->
                      let s' = removeClient client s in return (s', s')
                  broadcast ("SYSTEM: " `mappend` fst client `mappend` " disconnected.") s

talk :: WS.Connection -> MVar ServerState -> Client -> IO ()
talk conn state (user, _) = forever $ do
  msg <- WS.receiveData conn
  case msg of
    _   | ("direct " `T.isPrefixOf` msg) && (length (words $ T.unpack msg) > 4) -> readMVar state >>= broadcast
        ("SYSTEM: " `mappend` user `mappend` " is so stupid that she or he cannot send proper direct message!😄")
        | "direct " `T.isPrefixOf` msg -> readMVar state >>= sendDirectMessage user msg
        | T.toLower msg == T.pack "ping" -> WS.sendTextData conn (T.pack "👻: pong")
        | T.unpack (T.toLower msg) =~ ("fuck" ::String) -> sendMessageWithVulgarism state user msg
        | otherwise -> readMVar state >>= broadcast
          (user `mappend` ": " `mappend` msg)
